### PR TITLE
Support https communication between nodes

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+.. Note: At this step only encryption is supported
+
+Encrypted Internal Communication Configuration
+----------------------------------------------
+
+Encryption is configured in `config.properties` file.
+Encryption properties are identical for all of the presto node roles (coordinator, worker)
+and must be set on every node in the cluster.
+
+Disable HTTP endpoint. Leaving HTTP endpoint enabled leads to a security hole.
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: put this in WARNING

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Specify node's domains. Every node must have it's own domain configured.
+Domains must be in the same parent domain to be able to easy generate custom SSL \
+certificates. e.g.: `coordinator.example.com`, `worker1.example.com`, `worker2.example.com`.
+
+.. code-block:: none
+
+    node.external-address=<node-domain>
+
+Generate Java Keystore File
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: are these sub-sections?

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      [no]:  yes
+
+    Enter key password for <presto>
+            (RETURN if same as keystore password):
+
+.. Note: Replace `example.com` with your actual top-level domain
+
+Keystore file must be distributed across the cluster.
+
+Enable HTTPS endpoint.
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: subsection?

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+.. code-block:: none
+
+    http-server.https.enabled=true
+    http-server.https.port=7778
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: is this port number configurable?

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+and must be set on every node in the cluster.
+
+Disable HTTP endpoint. Leaving HTTP endpoint enabled leads to a security hole.
+
+.. code-block:: none
+
+    http-server.http.enabled=false
+
+Specify node's domains. Every node must have it's own domain configured.
+Domains must be in the same parent domain to be able to easy generate custom SSL \
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: Is this in node.properties?

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            (RETURN if same as keystore password):
+
+.. Note: Replace `example.com` with your actual top-level domain
+
+Keystore file must be distributed across the cluster.
+
+Enable HTTPS endpoint.
+
+.. code-block:: none
+
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: Which file to add this to?

File: presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            this.configuration = configuration;
+        }
+
+        @Override
+        public <T> void onProvision(ProvisionInvocation<T> provision)
+        {
+            HttpClientConfig config = checkType(provision.provision(), HttpClientConfig.class, "config");
+            checkState(config.getKeyStorePath() == null);
+            checkState(config.getKeyStorePath() == null);
```
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java]: should be getKeyStorePassword

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java]: Nice catch, thanks

File: presto-docs/src/main/sphinx/security/internal-communication.rst (hunk 2 above)
Comment (arhimondr) [reply presto-docs/src/main/sphinx/security/internal-communication.rst]: I'm going to make them bullet points. All these steps are part of the same guideline.

File: presto-docs/src/main/sphinx/security/internal-communication.rst (hunk 3 above)
Comment (arhimondr) [reply presto-docs/src/main/sphinx/security/internal-communication.rst]: ditto

File: presto-docs/src/main/sphinx/security/internal-communication.rst (hunk 5 above)
Comment (arhimondr) [reply presto-docs/src/main/sphinx/security/internal-communication.rst]: For the simplicity i suggest above to modify only `config.properties`. We can implement auto-domain assignment in `presto-admin`, and modify `node.properties` there.

File: presto-docs/src/main/sphinx/security/internal-communication.rst (hunk 6 above)
Comment (arhimondr) [reply presto-docs/src/main/sphinx/security/internal-communication.rst]: All the properties must be added to the `config.properies` files, as it is said in section above.

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    http-server.http.enabled=false
+
+.. warning::
+
+    Leaving HTTP endpoint enabled leads to a security hole.
+
+2. Specify node's domains. Every node must have it's own domain configured.
```
Comment (ebd2) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: Do you mean domain or FQDN here? Domain implies a container for hosts, whereas coordinator.example.com is a host, not a container.

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+Domains must be in the same parent domain to be able to easy generate custom SSL
```
Comment (ebd2) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: s/easy/easily/

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+3. Generate Java Keystore File.
+
+.. code-block:: none
+
+    keytool -genkeypair -alias example.com -keyalg RSA -keystore keystore.jks
+    Enter keystore password:
+    Re-enter new password:
+    What is your first and last name?
+      [Unknown]:  *.example.com
```
Comment (ebd2) [inline presto-docs/src/main/sphinx/security/internal-communication.rst:50]: Is this literally '*', or should the \* be replaced by the hostname you're generating a certificate for?

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      [Unknown]:
+    What is the name of your organization?
+      [Unknown]:
+    What is the name of your City or Locality?
+      [Unknown]:
+    What is the name of your State or Province?
+      [Unknown]:
+    What is the two-letter country code for this unit?
+      [Unknown]:
+    Is CN=eiger, OU=Unknown, O=Unknown, L=Unknown, ST=Unknown, C=Unknown correct?
```
Comment (ebd2) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: CN should match the value from the firstname/lastname question

File: presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+            this.configuration = configuration;
+        }
+
+        @Override
+        public <T> void onProvision(ProvisionInvocation<T> provision)
+        {
+            HttpClientConfig config = checkType(provision.provision(), HttpClientConfig.class, "config");
+            checkState(config.getKeyStorePath() == null);
+            checkState(config.getKeyStorePassword() == null);
+            config.setKeyStorePath(configuration.getKeyStorePath());
```
Comment (ebd2) [inline presto-main/src/main/java/com/facebook/presto/server/InternalHttpClientConfigurationModule.java]: Will this fail on startup if the keystore doesn't exist/isn't readable? Fail fast would be highly desirable.

**Review by ebd2 — CHANGES_REQUESTED** [review]

File: presto-docs/src/main/sphinx/security/internal-communication.rst (hunk 10 above)
Comment (arhimondr) [reply presto-docs/src/main/sphinx/security/internal-communication.rst:50]: It is literally. Every presto node must be able to connect to any other node within the same cluster. We can create private certificates for every single node for the exact domain, and than specify it for server, and create a keystore that has all the public keys for all the domains, and specify it for client. But for the simplicity we suggest to just create one single keystore for the base domain with `*`, and distribute it across the cluster.

Comment (ebd2) [reply presto-docs/src/main/sphinx/security/internal-communication.rst:50]: I think you should discuss those two options between "3. Generate..." and the code block. What you've written already should work if you change the we's to you's.

Comment (arhimondr) [reply presto-docs/src/main/sphinx/security/internal-communication.rst:50]: @ebd2 done

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+Hosts must be in the same parent domain to be able to easily generate custom SSL
+certificates. e.g.: `coordinator.example.com`, `worker1.example.com`, `worker2.example.com`.
+
+.. code-block:: none
+
+    node.external-address=<node-domain>
+
+3. Generate Java Keystore File. Every Presto node must be able to connect to any other node within
+the same cluster. It is possible to create private certificates for every single node for the
+exact domain, and than specify it for the server (`http-server.https.keystore.path`), and create a
```
Comment (ebd2) [inline presto-docs/src/main/sphinx/security/internal-communication.rst:39]: s/than/then/

Comment (ebd2) [reply presto-docs/src/main/sphinx/security/internal-communication.rst:39]: Actually, this should probably be: It is possible..., specify it for the server..., create keystore..., and specify it....

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+keystore that contains all the public keys for all the domains, and specify it for the
+client (`http-client.https.keystore.path`). But for simplicity single keystore for the base domain
```
Comment (ebd2) [inline presto-docs/src/main/sphinx/security/internal-communication.rst:41]: Suggest: "For simplicity, you can create a single keystore for the base domain with a wildcard"

**Comment by ebd2** [comment]
LGTM % grammar in the docs. If you think the code is final, I can edit for you if you want.

**Comment by arhimondr** [comment]
> If you think the code is final

Yes, it is.

> I can edit for you if you want

That would be much appreciated

**Pull request closed without merging**